iree/compiler/Codegen/LLVMCPU/VectorContractCustomKernels.cpp

```diff
@@ -0,0 +1,317 @@
+#include "mlir/Conversion/VectorToLLVM/ConvertVectorToLLVM.h"
+#include "mlir/Dialect/LLVMIR/LLVMDialect.h"
+#include "mlir/Dialect/LLVMIR/LLVMTypes.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/Dialect/Vector/VectorOps.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir {
+namespace iree_compiler {
+
+namespace {
+
+// Returns true if `contractionOp` is of the form
+//   matrix * transposed_matrix.
+// That is, if there are 2 parallel iterators, say M and N, 1 additive reduction
+// iterator, say K, and the indexing maps are {{M, K}, {N, K}, {M, N}}.
+static bool isMatrixTimesMatrixTransposed(vector::ContractionOp contractionOp) {
+  // Check that the reduction is additive.
+  if (contractionOp.kind() != vector::CombiningKind::ADD) {
+    return false;
+  }
+  // Check that there are 2 parallel and 1 reduction iterators.
+  auto iteratorTypes = contractionOp.iterator_types().getValue();
+  if (iteratorTypes.size() != 3) {
+    return false;
+  }
+  SmallVector<int, 3> parallel_iterators;
+  SmallVector<int, 3> reduction_iterators;
+  for (int i = 0; i < 3; i++) {
+    if (isParallelIterator(iteratorTypes[i])) {
+      parallel_iterators.push_back(i);
+    } else if (isReductionIterator(iteratorTypes[i])) {
+      reduction_iterators.push_back(i);
+    } else {
+      return false;
+    }
+  }
+  if (parallel_iterators.size() != 2 || reduction_iterators.size() != 1) {
+    return false;
+  }
+  // Give the found iterators some idiomatic names.
+  const int M = parallel_iterators[0];
+  const int N = parallel_iterators[1];
+  const int K = reduction_iterators[0];
+  // Check that there are 3 indexing maps.
+  auto indexingMaps = contractionOp.indexing_maps().getValue();
+  if (indexingMaps.size() != 3) {
+    return false;
+  }
+  // Check that the indexing maps have the expected form.
+  const int expectedMapResults[3][2] = {{M, K}, {N, K}, {M, N}};
+  for (int m = 0; m < 3; ++m) {
+    auto map = indexingMaps[m].cast<AffineMapAttr>().getValue();
+    if (map.getNumDims() != 3 || map.getNumResults() != 2) {
+      return false;
+    }
+    for (int r = 0; r < 2; ++r) {
+      int actualMapResult =
+          map.getResults()[r].cast<AffineDimExpr>().getPosition();
+      if (actualMapResult != expectedMapResults[m][r]) {
+        return false;
+      }
+    }
+  }
+  return true;
+}
+
+// Returns true if `contractionOp` is of the form
+//   matrix * transposed_matrix
+// where matrix is a vector<{mSize}x{kSize}xType>, and
+// transposed_matrix is a vector<{nSize}x{kSize}xType>
+static bool isMatrixTimesMatrixTransposed(vector::ContractionOp contractionOp,
+                                          int64_t mSize, int64_t kSize,
+                                          int64_t nSize) {
+  if (!isMatrixTimesMatrixTransposed(contractionOp)) {
+    return false;
+  }
+  VectorType lhsType = contractionOp.lhs().getType().cast<VectorType>();
```

Comment:
I dont think you need to check this. IIUC the correct affine maps automatically indicate that the shapes are consistent too. So its part of op semantics.

Reply:
Yes, what Geoffrey says. I renamed this function to make that clearer: isMatrixTimesMatrixTransposed*OfGivenShape*